tests/test_issues/test_547.py

```diff
@@ -0,0 +1,49 @@
+import pytest
+
+import magma as m
+
+def test_basic():
+    class Test(m.Circuit):
+        IO = ["I", m.In(m.Bits[1]), "O", m.Out(m.Bits[1])]
+
+        @classmethod
+        def definition(io):
+            io.O[0] @= io.I[0]
+
+def test_slice():
+    class Test(m.Circuit):
+        IO = ["I", m.In(m.Bits[2]), "O", m.Out(m.Bits[2])]
+
+        @classmethod
+        def definition(io):
+            io.O[:] @= io.I[:]
+
+def test_compound():
+    class Test(m.Circuit):
+        IO = ["I0", m.In(m.Bits[1]), "I1", m.In(m.Bits[1]), "O", m.Out(m.Bits[2])]
+
+        @classmethod
+        def definition(io):
+            io.O[0] @= io.I0
+            io.O[1] @= io.I1
+
+def test_recursive():
+    class Test(m.Circuit):
+        IO = ["I", m.In(m.Array[3, m.Bits[2]]), "O", m.Out(m.Array[3, m.Bits[2]])]
+
+        @classmethod
+        def definition(io):
+            io.O[0:1][:][0] @= io.I[0:1][:][0] # [0]
+            io.O[1:2][0][0:1] @= io.I[1:2][0][0:1] # [1][0]
+            io.O[1][1:2] @= io.I[1][1:2] # [1][1]
+            io.O[2][:] @= io.I[2][:]
+
+def test_errors():
+    with pytest.raises(ValueError):
```

Comment:
If the above code is changed to use log rather than exception, than this would change. See https://github.com/phanrahan/magma/blob/7bcc06cd5d520798ff5d34b4c499687db3d38592/tests/test_wire/test_errors.py#L4 for usage of getting logger output using `caplog`.

Reply:
Probably not the right place to raise this issue but...
The way you are using caplog is testing the implementation not the interface.  Which is decidedly not agile. Unless we are saying the exact error message is part of the interface (seems dubious).

Reply:
You could just regex or match on the error string prefix if you want to avoid testing the exact message.  E.g just check for the presence of ERROR

Reply:
Agreed. We can do `assert caplog.records[0].levelname == "ERROR"`. But either way, hard to test a "wrong" program errors as expected unless exceptions are standard in your (rather than logging; open to this too).

Reply:
Personally I have don't really think it matters if it logs or errors, if the standard is logs I am fine with that.

I do prefer checking the `levelname == "ERROR"` to checking exact error messages.  It seems like the property we actually care about.